### PR TITLE
[Refactoring] convert hosting/*.js to typescript code

### PR DESCRIPTION
**Pull request by iida-hayato**
- [x] convert to typescript code
  - serve/hosting.ts
  - hosting/initMiddleware
  - hosting/normalizedHostingConfigs
  - hosting/implicitInit
- [x] check npm t
- [x] check npm run lint
- [x] local test

### CODE REVIEW

File: src/hosting/normalizedHostingConfigs.ts [hunk 1]
```diff
@@ -1,31 +1,38 @@
-const _ = require("lodash");
+import * as _ from "lodash";
 
-function _filterOnly(configs, onlyString) {
+type HostingConfig = { site: string; target: string };
```
Comment (samtstern) [inline src/hosting/normalizedHostingConfigs.ts]: Any reason to use a `type` here but use an `interface` for `TemplateServerResponse`?

File: src/hosting/normalizedHostingConfigs.ts [hunk 2]
```diff
@@ -1,31 +1,38 @@
...
-  });
+  return configs.filter((config: HostingConfig) =>
+    _.includes(onlyTargets, config.target || config.site)
+  );
 }
 
-module.exports = function(options) {
+/**
+ * @param options
+ * @return {any[] | HostingConfig[]}
```
Comment (samtstern) [inline src/hosting/normalizedHostingConfigs.ts]: This JSDoc doesn't match the signature (also I'd say we should just remove the doc comment dince it doesn't tell you anything the function signature doesn't show)

File: src/serve/hosting.ts [hunk 3]
```diff
@@ -1,22 +1,19 @@
-"use strict";
+import clc = require("cli-color");
+import { FirebaseError } from "../error";
+import * as utils from "../utils";
+import * as detectProjectRoot from "../detectProjectRoot";
+import functionsProxy from "../hosting/functionsProxy";
+import cloudRunProxy from "../hosting/cloudRunProxy";
+const superstatic = require("superstatic").server;
```
Comment (samtstern) [inline src/serve/hosting.ts]: Could this be `import { server } = require("superstatic")`

Note: I always get confused about imports so I could be totally wrong.

File: src/serve/hosting.ts [hunk 4]
```diff
@@ -1,22 +1,19 @@
...
-var functionsProxy = require("../hosting/functionsProxy").default;
-var cloudRunProxy = require("../hosting/cloudRunProxy").default;
-var normalizedHostingConfigs = require("../hosting/normalizedHostingConfigs");
-
-var MAX_PORT_ATTEMPTS = 10;
-var _attempts = 0;
-var server;
-
-function _startServer(options, config, port, init) {
+function _startServer(options: any, config: any, port: number, init: TemplateServerResponce): void {
```
Comment (samtstern) [inline src/serve/hosting.ts]: Should `config` here use the `HostingConfig` type from the other file?  If so should we move that somewhere more central?

**Comment by samtstern** [comment]
@iida-hayato thanks for doing this!  I left some small comments, hopefully @bkendall can do a review as well.

File: src/hosting/normalizedHostingConfigs.ts (hunk 1 above)
Comment (iida-hayato) [reply src/hosting/normalizedHostingConfigs.ts]: it's no reason. 
it seems more better to interface.

File: src/hosting/normalizedHostingConfigs.ts (hunk 2 above)
Comment (iida-hayato) [reply src/hosting/normalizedHostingConfigs.ts]: ok i'll remove this.
(it generate by lint --fix)

File: src/serve/hosting.ts (hunk 3 above)
Comment (iida-hayato) [reply src/serve/hosting.ts]: sorry.'`import { server } = require("superstatic")` cause build error.

I'll read the superstatic implementation 🤔

File: src/serve/hosting.ts (hunk 4 above)
Comment (iida-hayato) [reply src/serve/hosting.ts]: > Should config here use the HostingConfig type from the other file?
Yeah. I'll change that code.

> If so should we move that somewhere more central?
maybe yes.
this `config` dependents supertest. so, this is  more better  if` @types/supertest` is exists 🤔

**Comment by iida-hayato** [comment]
There seems to be a better name for the added types.
such as
- TemplateServerResponce
- HostingConfig
- ApplicationHandler

But i have no idea . 🤔

File: src/hosting/implicitInit.ts [hunk 5]
```diff
@@ -1,15 +1,23 @@
...
 
-var INIT_TEMPLATE = fs.readFileSync(__dirname + "/../../templates/hosting/init.js", "utf8");
+export interface TemplateServerResponce {
+  js: string | undefined;
+  json: string | undefined;
+}
 
-module.exports = async function(options) {
+/**
+ * generate template server responce
```
Comment (bkendall) [inline src/hosting/implicitInit.ts]:
```suggestion
 * Generate template server response.
```

"response" is spelled incorrectly here and in the type. Please fix.

File: src/hosting/implicitInit.ts [hunk 6]
```diff
@@ -1,15 +1,23 @@
...
+ * @param options
+ * @return {Promise<{js: string, json: string}>}
```
Comment (bkendall) [inline src/hosting/implicitInit.ts]: Specifying return types in a typescript file is redundant. Please just include a brief written description of what is returned. e.g.:

```suggestion
 * @return the initialization configuration.
```

File: src/hosting/normalizedHostingConfigs.ts (hunk 1 above)
Comment (bkendall) [reply src/hosting/normalizedHostingConfigs.ts]: please prefer interfaces to types. Maybe that's something I could enforce with the linter in the future...

File: src/hosting/normalizedHostingConfigs.ts [hunk 7]
```diff
@@ -1,31 +1,38 @@
...
 
-function _filterOnly(configs, onlyString) {
+type HostingConfig = { site: string; target: string };
+
+/**
+ * @param configs
+ * @param onlyString
+ * @return {HostingConfig[]}
+ */
+function _filterOnly(configs: HostingConfig[], onlyString: string): HostingConfig[] {
```
Comment (bkendall) [inline src/hosting/normalizedHostingConfigs.ts]: unexported (private) functions do not need to be prefixed with an underscore.

File: src/hosting/normalizedHostingConfigs.ts (hunk 2 above)
Comment (bkendall) [reply src/hosting/normalizedHostingConfigs.ts]: `lint --fix` adds this return type? I will dig into that as it shouldn't in TS...

File: src/serve/hosting.ts [hunk 8]
```diff
@@ -1,22 +1,19 @@
...
+import cloudRunProxy from "../hosting/cloudRunProxy";
+const superstatic = require("superstatic").server;
+import { implicitInit, TemplateServerResponce } from "../hosting/implicitInit";
+import { initMiddleware } from "../hosting/initMiddleware";
+import { normalizedHostingConfigs } from "../hosting/normalizedHostingConfigs";
 
-var clc = require("cli-color");
-var { FirebaseError } = require("../error");
+const MAX_PORT_ATTEMPTS = 10;
+let _attempts = 0;
```
Comment (bkendall) [inline src/serve/hosting.ts]: To keep with ts convention: any private (not exported) functions and variables should not be prefaced with an underscore.

File: src/serve/hosting.ts [hunk 9]
```diff
@@ -67,31 +64,25 @@ function _startServer(options, config, port, init) {
   });
 }
 
-function _stop() {
+export function stop(): Promise<void> {
```
Comment (bkendall) [inline src/serve/hosting.ts]: `export async function`? same blow for start and connect?

**Comment by iida-hayato** [comment]
Generated JSDocs seems useless,  so I erased and rewrite JSDoc comments.
Tell me if I wrote wrong comments.

**Comment by bkendall** [comment]
FYI - I've still got this on my radar, but dedicating the time to make sure I can go through it is a little tricky. I'm hoping to get back to you today or tomorrow. Thanks for the PR and your patience.

**Comment by iida-hayato** [comment]
@bkendall 

All right.
today, I fixed code  as I can.

If this still difficult to merge, you can close it.

or, Is there any other code that  preferentially converts to typescript?
I can help that.

**Comment by bkendall** [comment]
Thanks for keeping on top of this. I've been horrible at actually getting back to you on this and I apologize for that - things have been rather busy for me as of late as other things have taken priority. I appreciate the PR and promise I haven't lost track of it!

**Comment by iida-hayato** [comment]
Because base's host.js has been changed,
It would be better to close this PR and recreate a new PR 🤔

**Comment by bkendall** [comment]
I've been super super busy, so I haven't gotten time to sit down on this, but I was going to manually resolve the conflicts and merge this. If you can manage to resolve the conflicts yourself in the next few days, I will promise to re-review by the end of the week. Otherwise, it may be next week when I get a chance to wrap this up.

**Comment by bkendall** [comment]
(seriously, thank you for your patience and I'm sorry I've been slooww)

**Comment by bkendall** [comment]
🎶 I'm horrible, it's true 🎶

(to the tune of You're Beautiful)

I'm looking at this right now!

**Pull request closed without merging**